File: application/synoAuth.php
<?php

error_reporting(E_ALL | E_STRICT);  
ini_set('display_startup_errors',1);  
ini_set('display_errors',1);

class SynoAuth {
	protected $synotoken;
	protected $username;
	protected $usergroups;
	protected $logged;
	protected $REMOTE_ADDR;
	protected $cookie_id;

	// Constructor
	public function SynoAuth()
	{
		$this->logged = False; // By default: not logged
		$this->REMOTE_ADDR = escapeshellcmd($_SERVER['REMOTE_ADDR']); // Get IP Address
		$this->cookie_id = escapeshellcmd($_COOKIE['id']); // Get Cookie ID
	}

	// Get login information
	// If token is provided, use it. If not, use the previously provided one (with setToken method)
	public function login($token="")
	{
		if ($token == "")
		{
			$token = $this->getToken();
		} else
		{
			$token = $this->setToken($token);
		}
		if ($token == "")
		{
			return False;
		}
		if ($this->_authentificate() == "")
			return False;
		$this->_getGroups();
		return $this->getUsername();
	}

	// Call authenticate.cgi with provided synoToken, IP address and Cookie ID in order to get the username
	protected function _authentificate()
	{
		$cmd = 'export QUERY_STRING="SynoToken=' . escapeshellcmd($this->synotoken).'" ';
		$cmd .= 'REMOTE_ADDR=' . escapeshellarg($this->REMOTE_ADDR) . ' ';
		$cmd .= 'HTTP_COOKIE="id=' . escapeshellcmd($this->cookie_id) . '" ';
		$cmd .= '&& /usr/syno/synoman/webman/authenticate.cgi';
		exec($cmd,$result);
		if (count($result) < 1)
		{
			return False;
		}
		$this->username = $result[0];
		return $this->getUsername();
	}

	// Get groups
	protected function _getGroups()
	{
		// Only if userdata has already been determined
		if (escapeshellarg($this->username) == "")
			return False;
		exec('id -Gn ' . escapeshellarg($this->username) . ' 2>/dev/null',$result);
		$this->usergroups = explode(" ",$result[0]);
		return $this->getGroups();
	}

	public function getGroups()
	{
		return $this->usergroups;
	}

	public function setToken($token)
	{
		$this->synotoken = escapeshellcmd($token);
		return $this->getToken();
	}

	public function getToken()
	{
		return escapeshellcmd($this->synotoken);
	}

	public function getUsername()
	{
		return escapeshellcmd($this->username);
	}
}
?>

File: application/synoAuth.py
#!/usr/bin/env python
#encoding:utf-8

import os
import sys
import cgi,cgitb
import Cookie
import subprocess

class SynoAuth():
	# Constructor
	def __init__(self):
		self.synotoken = ""
		self.username = ""
		self.usergroups = ""
		self.logged = False # By default: not logged
		if "REMOTE_ADDR" in os.environ:
			self.REMOTE_ADDR = os.environ["REMOTE_ADDR"] # Get IP Address
		else:
			self.REMOTE_ADDR = ""
		if 'HTTP_COOKIE' in os.environ: # Get Cookie ID
			cookie_string=os.environ.get('HTTP_COOKIE')
			c=Cookie.SimpleCookie()
			c.load(cookie_string)
			try:
				self.cookie_id=c['id'].value
			except KeyError:
				self.cookie_id=''
		else:
			self.cookie_id=''

	''' Get login information
	If token is provided, use it. If not, use the previously provided one (with setToken method) '''
	def login(self,token=""):
		if token == "":
			token = self.getToken()
		else:
			token = self.setToken(token)

		if token == "":
			return False

		if self._authentificate() == "":
			return False
		self._getGroups()
		return self.getUsername()

	# Call authenticate.cgi with provided synoToken, IP address and Cookie ID in order to get the username
	def _authentificate(self):
		os.environ['QUERY_STRING'] = 'SynoToken=' + self.synotoken.replace("'", "'\\''")
		#os.environ['REMOTE_ADDR'] = self.REMOTE_ADDR,
		os.environ['HTTP_COOKIE'] = 'id=' + self.cookie_id
		cmd = ['/usr/syno/synoman/webman/authenticate.cgi']
		try:
			result = subprocess.check_output(cmd).replace('\n','')
		except:
			result = ""
		if len(result) < 1:
			return False

		self.username = result.replace("\n","")
		return self.getUsername()

	# Get groups
	def _getGroups(self):
		# Only if userdata has already been determined
		if self.username.replace("'", "'\\''") == "":
			return False
		cmd = ['id','-Gn',self.username.replace("'", "\\'")]
		try:
			result = subprocess.check_output(cmd)
		except:
			result = ""
		self.usergroups = result.replace("\n","").split(" ")
		return self.getGroups()

	def getGroups(self):
		return self.usergroups

	def setToken(self,token):
		try:
			self.synotoken = token.replace("'","\\'")
		except:
			print token
			sys.exit()
		return self.getToken()

	def getToken(self):
		return self.synotoken.replace("'","\\'")

	def getUsername(self):
		return self.username.replace("'","\\'")

